common/src/main/java/com/microsoft/identity/common/logging/Logger.java

```diff
@@ -0,0 +1,575 @@
+                     final boolean containsPII) {
+
+        final String dateTimeStamp = getUTCDateTimeAsString();
+
+        sLogExecutor.execute(new Runnable() {
+            @Override
+            public void run() {
+                if (logLevel.compareTo(mLogLevel) > 0) {
+                    return;
+                }
+
+                // Developer turns off PII logging, if the log meLoggerSettingssage contains any PII,
+                // we should not send it.
+                if (!sAllowPii && containsPII) {
+                    return;
+                }
+
+                //Format the log message.
+                final String logMessage = formatMessage(correlationID, message, dateTimeStamp, throwable);
+
+                // Send logs into Logcat.
+                if (sAllowLogcat) {
+                    sendLogcatLogs(tag, logLevel, logMessage);
+                }
+
+                // Send logs into external logger callback.
+                final ILoggerCallback externalLogger = mExternalLogger;
+
+                if (null != externalLogger) {
+                    try {
+                        mExternalLogger.log(tag, logLevel, logMessage, containsPII);
+                    } catch (final Exception e) {
+                        // log message as warning to report callback error issue
+                        if (!containsPII || sAllowPii) {
+                            Log.w(tag, String.format(CUSTOM_LOG_ERROR, logMessage));
+                        }
+                    }
+                }
+            }
+        });
+    }
+
+    /**
+     * Wrap the log message.
+     * If correlation id exists:
+     * <library_version> <platform> <platform_version> [<timestamp> - <correlation_id>] <log_message>
+     * If correlation id doesn't exist:
+     * <library_version> <platform> <platform_version> [<timestamp>] <log_message>
+     */
+    private String formatMessage(@Nullable final String correlationID,
+                                 @Nullable final String message,
+                                 @NonNull final String dateTimeStamp,
+                                 @Nullable final Throwable throwable) {
+        final String logMessage = StringExtensions.isNullOrBlank(message) ? "N/A" : message;
+        return " [" + dateTimeStamp
+                + (StringExtensions.isNullOrBlank(correlationID) ? "] " : " - " + correlationID + "] ")
+                + logMessage
+                + " Android " + Build.VERSION.SDK_INT
+                + (throwable == null ? "" : '\n' + Log.getStackTraceString(throwable));
+    }
+
+    private static String getUTCDateTimeAsString() {
+        final SimpleDateFormat dateFormat = new SimpleDateFormat(DATE_FORMAT, Locale.getDefault());
+        dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+
+        return dateFormat.format(new Date());
+    }
+
+    /**
+     * Send logs to logcat as the default logging if developer doesn't turn off the logcat logging.
```

Comment:
I think this just send logs to logcat regardless of whether developer turns it on or off. This method doesn't seem to care....it seems the caller needs to take care of that...we can probably update this javadoc

Reply:
0ffd5d6085fa623073dfcebefa57259473648e65